numpy/core/src/multiarray/ctors.c

```diff
@@ -1701,6 +1701,15 @@ PyArray_FromAny(PyObject *op, PyArray_Descr *newtype, int min_depth,
                     (dtype == NULL) ? PyArray_DESCR(arr) : dtype,
                     &newtype);
     }
+    
+    /* if object was not requested, but we got it, and is not 0d, fail */
+    if (ndim != 0 && dtype != NULL && dtype->type_num == NPY_OBJECT) {
```

Comment:
In the end this block was the entire implementation. To disallow non-explicit 0d-object-arrays remove the `ndim!= 0` check.